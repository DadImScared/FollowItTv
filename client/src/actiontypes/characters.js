
export const ADD_CHARACTERS = 'characters/ADD_CHARACTERS';
