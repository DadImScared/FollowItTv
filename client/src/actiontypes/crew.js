
export const ADD_CREW = 'crew/ADD_CREW';
