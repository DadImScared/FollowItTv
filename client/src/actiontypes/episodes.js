
export const ADD_EPISODES = 'episodes/ADD_EPISODES';
