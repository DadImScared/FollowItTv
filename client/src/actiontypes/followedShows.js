
export const FOLLOW_SHOW = 'followedShows/FOLLOW_SHOW';
export const UNFOLLOW_SHOW = 'followedShows/UNFOLLOW_SHOW';
export const ADD_DAY = 'followedShows/ADD_DAY';
