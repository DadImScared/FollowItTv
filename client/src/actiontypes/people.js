
export const ADD_PEOPLE = 'people/ADD_PEOPLE';
