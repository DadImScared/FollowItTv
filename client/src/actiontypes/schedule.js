
export const ADD_SCHEDULE = 'schedule/ADD_SCHEDULE';
