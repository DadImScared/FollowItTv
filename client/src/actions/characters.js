
import { ADD_CHARACTERS } from '../actiontypes/characters';
import { loadingEnd, loadingStart } from './loading';
import { castSchema } from '../schemas/person';
import { normalize } from 'normalizr';
import { baseUrl } from './seasons';
import axios from 'axios';

export const addCharacters = (entities) => ({
  type: ADD_CHARACTERS,
  ...entities
});

const getCast = async (showId) => {
  return await axios.get(`${baseUrl}shows/${showId}/cast`);
};

export const handleCastData = (data) => {
  const { entities: { person, character }, result } = normalize(data, castSchema);
  const characterIds = result.reduce((acc, { character }) => {
    const newAcc = [...acc];
    if (!newAcc.includes(character)) {
      newAcc.push(character);
    }
    return newAcc;
  }, []);
  return {
    person,
    character,
    result,
    characterIds
  };
};

export const requestCast = (showId) => {
  return async (dispatch) => {
    const loadingId = `GET_CAST_${showId}`;
    dispatch(loadingStart(loadingId));
    try {
      const { data } = await getCast(showId);
      const { person: people, character: characters, characterIds } = handleCastData(data);
      dispatch(addCharacters({ id: showId, people, characters, characterIds }));
      dispatch(loadingEnd(loadingId));
    }
    catch (e) {
      dispatch(loadingEnd(loadingId));
      throw e;
    }
  };
};
