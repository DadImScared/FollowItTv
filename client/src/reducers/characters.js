
import { ADD_CHARACTERS } from '../actiontypes/characters';

function characters(state = {}, action) {
  switch(action.type) {
    case ADD_CHARACTERS:
      return {
        ...state,
        ...action.characters
      };
    default:
      return state;
  }
}

export default characters;
