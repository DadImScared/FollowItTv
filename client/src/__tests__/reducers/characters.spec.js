
import { ADD_CHARACTERS } from '../../actiontypes/characters';
import reducer from '../../reducers/characters';


describe('characters reducer', () => {
  it('should add characters', () => {
    expect(
      reducer({}, {
        type: ADD_CHARACTERS,
        characters: { 1: { id: 1 } }
      })
    ).toMatchSnapshot();
  });
});
