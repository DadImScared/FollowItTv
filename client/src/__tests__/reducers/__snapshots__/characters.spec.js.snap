// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`characters reducer should add characters 1`] = `
Object {
  "1": Object {
    "id": 1,
  },
}
`;
